agent/lib_composer.c

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "php_agent.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "nr_txn.h"
+#include "util_logging.h"
+#include "util_memory.h"
+#include "util_syscalls.h"
+
+static bool nr_execute_handle_autoload_composer_is_initialized() {
+  zend_class_entry* zce = NULL;
+
+  if (NULL == (zce = nr_php_find_class("composer\\installedversions"))) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "Composer\\InstalledVersions class not found");
+    return false;
+  };
+
+  // the class is found - there's hope!
+  if (NULL == nr_php_find_class_method(zce, "getinstalledpackages")
+      || NULL == nr_php_find_class_method(zce, "getversion")) {
+    nrl_verbosedebug(
+        NRL_INSTRUMENT,
+        "Composer\\InstalledVersions class found, but methods not found");
+    return false;
+  }
+
+  return true;
+}
+
+static int nr_execute_handle_autoload_composer_init(const char* vendor_path) {
+  char* code = NULL;
+  zval retval;
+  int result = -1;
+
+  if (nr_execute_handle_autoload_composer_is_initialized()) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: already initialized", __func__);
+    return NR_SUCCESS;
+  }
+
+#if 0
+  (void)vendor_path;
+  code
+      = nr_formatf(""
+        "(function() {"
+        "  try {"
+        "      if (class_exists('Composer\\InstalledVersions')) {"
+        "          if (method_exists('Composer\\InstalledVersions', "
+        "             'getInstalledPackages') && method_exists('Composer\\InstalledVersions', "
+        "             'getVersion')) {"
+        "            return true;"
+        "          } else {"
+        "            return false;"
+        "          }"
+        "      } else {"
+        "        return false;"
+        "      }"
+        "  } catch (Exception $e) {"
+        "      return NULL;"
+        "  }"
+        "})();");
+#else
+  code = nr_formatf("include_once '%s/composer/InstalledVersions.php';",
+                    vendor_path);
+#endif
+
+  result = zend_eval_string(code, &retval, "newrelic\\init_composer_api");
+  if (result != SUCCESS) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: zend_eval_string(%s) failed, result=%d", __func__,
+                     code, result);
+    return NR_FAILURE;
```

Comment:
If we return here, code doesn't get freed.
I'm not sure what the state of retval is and if it also needs a zval_dtor if the eval_string fails.

Reply:
Thanks for pointing it out. Memory leak is fixed in 060303ef42268a0f5d4949bbd3df42d800dd650b. I need to take a closer look at `retval`.

Reply:
https://github.com/newrelic/newrelic-php-agent/pull/962/commits/aab8068fb61792807fce5eac658bb126ee314223 changes how the agent interacts with Composer's runtime API - there's no longer need to allocate memory for the PHP code to be executed, because the code string is static. This patch is currently under test, but if it passes it will become the final version.

As far as `retval` goes `zend_eval_string` leaves it in good state as long as it returns `SUCCESS`. Once that happens, `zval_dtor` can be called. https://github.com/newrelic/newrelic-php-agent/pull/962/commits/aab8068fb61792807fce5eac658bb126ee314223 incorporates this approach.